### PR TITLE
feat(declarative) ability to run Kong without a db and configured declaratively

### PR DESCRIPTION
**Pull request by hutchic**

### CODE REVIEW

File: kong-ingress-data-plane-postgres.yaml [hunk 1]
```diff
@@ -75,5 +79,13 @@ spec:
     port: 443
     targetPort: 8443
     protocol: TCP
+  - name: http
+    port: 80
```
Comment (hbagdi) [inline kong-ingress-data-plane-postgres.yaml]: Port 80 is being bound to multiple target ports, once for Proxy side and once for Admin API, that doesn't look right?

File: kong-ingress-data-plane-postgres.yaml [hunk 2]
```diff
@@ -75,5 +79,13 @@ spec:
...
+    targetPort: 8001
+    protocol: TCP
+  - name: https
+    port: 443
```
Comment (hbagdi) [inline kong-ingress-data-plane-postgres.yaml]: Same here for 443.

File: kong-dbless.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  type: NodePort
+  ports:
+  - name: kong-proxy-http
+    port: 8000
+    targetPort: 8000
+    protocol: TCP
+  - name: kong-proxy-https
+    port: 8443
+    targetPort: 8443
+    protocol: TCP
```
Comment (hbagdi) [inline kong-dbless.yaml]: If the service is named `kong-control-plane`, should it be able to route proxy traffic?

File: kong-dbless.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+kind: Deployment
+metadata:
+  namespace: kong
+  name: kong-control-plane
+  labels:
+    app: kong-control-plane
+spec:
+  selector:
+    matchLabels:
+      app: kong-control-plane
```
Comment (hbagdi) [inline kong-dbless.yaml]: This name doesn't sound right. Kong in DB-less mode is merely a proxy with a bunch of config. Should this be names something simpler?

File: README.md [hunk 5]
```diff
@@ -44,6 +44,10 @@ export PROXY_PORT=$(kubectl get svc --namespace kong kong-ingress-data-plane -o
 curl $HOST:$PROXY_PORT
 ```
 
+If using dbless/declarative the `declarative.yaml` file is mounted as a config map onto the Kong containers. We use the md5sum of `declarative.yaml`
+file to update the deployment but per [Facilitate ConfigMap rollouts / management](https://github.com/kubernetes/kubernetes/issues/22368) for production
+setups one would might be best to use helm, kustomize or [reloader](https://github.com/stakater/reloader).
+
```
Comment (hbagdi) [inline README.md]: Please respect 80 column length in this file.

File: kong-dbless.yaml (hunk 4 above)
Comment (hutchic) [reply kong-dbless.yaml]: I wasn't sure what to call it figured it'd get debated on the PR ;)

Is just `kong` sufficient I wonder?

File: kong-ingress-data-plane-postgres.yaml (hunk 1 above)
Comment (hutchic) [reply kong-ingress-data-plane-postgres.yaml]: agreed fixed on the parent branch f0d9f53be25e49a1c1cef1b84eb5b1a9014e137a

File: kong-ingress-data-plane-postgres.yaml (hunk 2 above)
Comment (hutchic) [reply kong-ingress-data-plane-postgres.yaml]: agreed fixed on the parent branch f0d9f53be25e49a1c1cef1b84eb5b1a9014e137a

File: README.md (hunk 5 above)
Comment (hutchic) [reply README.md]: b72b3d0 does this 👍

File: kong-dbless.yaml [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    targetPort: 8443
+    protocol: TCP
+  - name: http
+    port: 8001
+    targetPort: 8001
+    protocol: TCP
+  - name: https
+    port: 8444
+    targetPort: 8444
+    protocol: TCP
```
Comment (hbagdi) [inline kong-dbless.yaml]: This is overloading the service with multiple concerns.
We should split this into two k8s services: kong-proxy and kong-admin.

File: Makefile [hunk 7]
```diff
@@ -1,25 +1,34 @@
...
-	-kubectl delete csr kong-control-plane.kong.svc
-	kubectl delete -f kong-namespace.yaml
+
+k8s_setup:
+	kubectl apply -f kong-namespace.yaml
+	-./setup_certificate.sh
+
+run_dbless: k8s_setup
+	-kubectl create configmap kongdeclarative -n kong --from-file=declarative.yaml
+	kubectl create configmap kongdeclarative -n kong --from-file=declarative.yaml -o yaml --dry-run | kubectl replace -n kong -f -
```
Comment (hbagdi) [inline Makefile]: Why is this needed?
Isn't a kubectl apply sufficient here?

File: Makefile [hunk 8]
```diff
@@ -1,25 +1,34 @@
...
+	kubectl apply -f kong-dbless.yaml
+	kubectl patch deployment kong-control-plane -n kong -p "{\"spec\":{\"template\":{\"metadata\":{\"labels\":{\"declarative\":\"`md5sum declarative.yaml | awk '{ print $$1 }'`\"}}}}}}"
```
Comment (hbagdi) [inline Makefile]: `s/kong-control-plane/kong-dbless` ?

File: kong-ingress-data-plane-postgres.yaml [hunk 9]
```diff
@@ -39,14 +39,18 @@ spec:
...
             value: 0.0.0.0:8000,0.0.0.0:8443 ssl
         ports:
         - name: data-http
           containerPort: 8000
         - name: data-https
           containerPort: 8443
+        - name: admin-http
+          containerPort: 8001
+        - name: admin-https
+          containerPort: 8444
```
Comment (hbagdi) [inline kong-ingress-data-plane-postgres.yaml]: Why is the Admin API being enabled on the data-plane Kong nodes?

File: Makefile (hunk 7 above)
Comment (hutchic) [reply Makefile]: I don't think it is no

```
kubectl apply configmap kongdeclarative -n kong --from-file=declarative.yaml
Error: unknown flag: --from-file
```

File: Makefile (hunk 8 above)
Comment (hutchic) [reply Makefile]: resolved ffe4489

File: kong-ingress-data-plane-postgres.yaml (hunk 9 above)
Comment (hutchic) [reply kong-ingress-data-plane-postgres.yaml]: oh weird I shouldn't have been modifying this file at all ... I'll fix 👍

File: kong-dbless.yaml (hunk 6 above)
Comment (hutchic) [reply kong-dbless.yaml]: I'm a little hazy as to what the separation of concerns should look like with dbless / declarative.

> We should split this into two k8s services: kong-proxy and kong-admin.
Counterpoint: dbless shouldn't have an admin api in which case should it just be turned off entirely?

Comment (hutchic) [reply kong-dbless.yaml]: discussed offline and agreed dbless shouldn't expose admin api at all 👍

File: Makefile (hunk 7 above)
Comment (hbagdi) [reply Makefile]: Oh right, my bad.
Second question, why is the create being done twice? It Line 7 seems a bit weird, a left-over?

Comment (hutchic) [reply Makefile]: Ah yeah it's kind of a hack (open to alternatives). It lets the make task succeed on first run and also let's one edit `declarative.yaml` and re-run the same task against an already running Kong dbless
